improver/calibration/rainforest_calibration.py

```diff
@@ -0,0 +1,170 @@
+
+# Passed to choose_set_of_percentiles to set of evenly spaced percentiles
+DEFAULT_ERROR_PERCENTILES_COUNT = 19
+DEFAULT_OUTPUT_REALIZATIONS_COUNT = 100
+
+
+class TrainRainForestsTreeModels(BasePlugin):
+    """Class to train tree models for use in RainForests calibration."""
+
+
+class ApplyRainForestsCalibration(PostProcessingPlugin):
+    """Class to calibrate input forecast given a series of RainForests tree models."""
+
+    def __init__(self, model_config_dict: dict, nthreads: int):
+        """Initialise the tree model variables used in the application of RainForests
+        Calibration.
+
+        Args:
+            model_config_dict:
+                Dictionary containing Rainforests model configuration variables.
+            nthreads:
+                Number of threads to use when initialising tree-model objects.
+
+        Dictionary is of format::
+
+            {
+                "-50.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
+                    "treelite_model" : "<path_to_treelite_model_object>"
+                },
+                "-25.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
+                    "treelite_model" : "<path_to_treelite_model_object>"
+                },
+                ...,
+                "50.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
+                    "treelite_model" : "<path_to_treelite_model_object>"
+                }
+            }
+
+        The keys specify the error threshold value, while the associated values
+        are the path to the corresponding tree-model objects for that threshold.
+
+        Treelite Predictors are returned if treelite_runtime is an installed dependency and
+        an associated path has been supplied for all thresholds, otherwise lightgbm the
+        Boosters are returned.
+
+        Returns:
+            - numpy array containing the error threshold values.
+            - list of tree-model objects to be used in calibration process.
+        """
+        import importlib
+
+        from lightgbm import Booster
+
+        if importlib.util.find_spec("treelite") is not None:
+            from treelite_runtime import Predictor
+
+            self.treelite_enabled = True
+        else:
+            warnings.warn(
+                "Module treelite_runtime unavailable. Defaulting to using lightgbm Boosters."
+            )
+            self.treelite_enabled = False
+
+        error_thresholds = list(model_config_dict.keys())
+
+        lightgbm_model_filenames = [
+            model_config_dict[threshold].get("lightgbm_model")
+            for threshold in error_thresholds
+        ]
+        treelite_model_filenames = [
+            model_config_dict[threshold].get("treelite_model")
+            for threshold in error_thresholds
+        ]
+        if (None not in treelite_model_filenames) and self.treelite_enabled:
```

Comment:
The default behaviour for `dict.get` is to return `None` when no default value is provided (https://docs.python.org/3/library/stdtypes.html#dict.get). I could update the `.get` calls to be more explicit, but I think as it stands the code should not produce a `KeyError`.

The intention here is that in order to use treelite `Predictor` for the tree_model, all error thresholds must have a path specified for `treelite_model` otherwise it will default to using lightgbm `Booster` instead. I believe the code does this currently.

However, looking over this I think the situation that has not been addressed if where a `treelite_model` is specified, but not the corresponding `lightgbm_model`. In this case the code will fail as a `None` value will be passed into the `Booster` initialisation. I can add a check to ensure there are no `None` values in `lightgbm_model_filenames` before loading to catch this situation.

Reply:
My mistake. I was wrong on the default behaviour of `get`. I'm now wondering if the behaviour has changed, or if I was always wrong! Never mind...

You are right in that the behaviour does not trap incomplete `lightgbm_model` file names, so the process should fail somewhere else. I don't know where it will fail as your tests don't cover this. I suggest you add a test for this, then modify the behaviour if necessary to return a useful Exception.

Reply:
No worries, it's always good to check these things.

As described in the next comment, I've added a check for this case and added a unit-test to check correct behaviour.